### DISCUSSION TITLE
[BUG]: Micros bugs out below ~370 microseconds

### DISCUSSION THREAD

**Discussion by Stipena** [jonas-merkle/AS5047P#28] [Closed issues]
**Describe the bug**
If state machine method is used with "Micros" to control steps for stepper motor, bug happens when step on and off intervals is set to lower than ~370 microseconds. Changing interval to 350 microseconds, motor suddenly starts spinning ~50-70% faster. If setting interval to 50 microseconds, motor speed is the same as it were set to 350.

**To Reproduce**
Arduino UNO R3, A4988 stepper driver, Bipolar stepper motor, AS5047P-TS_EK_AB encoder.
Windows 10 x64, Arduino IDE 1.8.12

**Expected behavior**
Motor should spin full speed. Full speed should be reached at 43 microseconds step interval.

 - Arduino Board Type [Uno]
 - Arduino Software Version [1.8.12]
 - The Version of this Library [2.1.2]

**Additional context**
In schematics there is LCD Keypad Shield, but it isn't reason of the problem since I tried code without it and had same issue.
Tried to use int instead of float to store angle position, it maybe lowered the threshold from ~370 to ~360 microseconds.
![Kembrik Pooper 3000v2](https://user-images.githubusercontent.com/4146267/113362930-39bc9f00-9358-11eb-9cc2-bf85f013f1ce.png)
```
#include <AS5047P.h>
#include <LiquidCrystal.h>
#include <LCDKeypad.h>
LCDKeypad lcd;

//initial backlight
unsigned char bckl = 255;

// define the chip select port.
#define AS5047P_CHIP_SELECT_PORT 3

// define the spi bus speed
#define AS5047P_CUSTOM_SPI_BUS_SPEED 100000

int LCDrefreshRate = 40;
unsigned long LCDmillis = 0;
unsigned long CutterStepInterval = 43;
unsigned long CutterStepMicros = 0;
unsigned long DirectionChangeMillis = 0;
int DirectionChangeInterval = 3000;

const int EnableCutterPin = 15;
const int CutterDirPin = 16;
const int CutterStepPin = 17;
int EnableCutterState = LOW; // Cutter Enabled
int CutterDirState = HIGH; // Cutter spin clockwise
int CutterStepState = LOW; // Cutter step off
int DirectionChange = 0;

// initialize a new AS5047P sensor object.
AS5047P as5047p(AS5047P_CHIP_SELECT_PORT, AS5047P_CUSTOM_SPI_BUS_SPEED);

float angle;

void setup() {
  pinMode(EnableCutterPin, OUTPUT); //Enable Cutter
  pinMode(CutterDirPin, OUTPUT); //Cutter Dir
  pinMode(CutterStepPin, OUTPUT); //Cutter Step
  pinMode(10, OUTPUT); //LCD backlight control pin
  analogWrite(10, bckl);
  lcd.begin(16, 2);
  lcd.clear();

  // initialize the AS5047P sensor and hold if sensor can't be initialized..
  while (!as5047p.initSPI()) {
    lcd.print(F("No AS5047P"));
    delay(5000);
  }
}

void loop() {
  unsigned long currentMillis = millis();
  unsigned long currentMicros = micros();
  angle = as5047p.readAngleDegree();

  // refresh LCD and display cutter encoder angle
  if (currentMillis - LCDmillis >= LCDrefreshRate) {
    LCDmillis = currentMillis;
    lcd.clear();
    lcd.print(angle);      // read the angle value from the AS5047P sensor an print it.
  }

  // Direction change
  if (angle >= 350.00 && DirectionChange == 0) {
    DirectionChange = 3;
    DirectionChangeMillis = currentMillis;
    digitalWrite(EnableCutterPin, HIGH);
  }
  if (currentMillis - DirectionChangeMillis >= DirectionChangeInterval && DirectionChange == 3) {
    DirectionChange = 1;
    digitalWrite(EnableCutterPin, LOW);
  }
  if (angle <= 10.00 && DirectionChange == 1) {
    DirectionChange = 4;
    DirectionChangeMillis = currentMillis;
    digitalWrite(EnableCutterPin, HIGH);
  }
  if (currentMillis - DirectionChangeMillis >= DirectionChangeInterval && DirectionChange == 4) {
    DirectionChange = 0;
    digitalWrite(EnableCutterPin, LOW);
  }

  // stepper control clockwise
  if (currentMicros - CutterStepMicros >= CutterStepInterval && angle <= 350.00 && DirectionChange == 0) {
    // save the last time you made step
    CutterStepMicros = currentMicros;
    CutterDirState = HIGH;
    digitalWrite(CutterDirPin, CutterDirState); // Cutter spin clockwise
    // if step is off turn it on and vice-versa:
    if (CutterStepState == LOW) {
      CutterStepState = HIGH;
    } else {
      CutterStepState = LOW;
    }
    digitalWrite(CutterStepPin, CutterStepState);
  }

  // stepper control counter-clockwise
  if (currentMicros - CutterStepMicros >= CutterStepInterval && angle >= 10.00 && DirectionChange == 1) {
    // save the last time you made step
    CutterStepMicros = currentMicros;
    CutterDirState = LOW;
    digitalWrite(CutterDirPin, CutterDirState); // Cutter spin counter-clockwise
    // if step is off turn it on and vice-versa:
    if (CutterStepState == LOW) {
      CutterStepState = HIGH;
    } else {
      CutterStepState = LOW;
    }
    digitalWrite(CutterStepPin, CutterStepState);
  }
}
```

**Comment by Stipena** [comment]
Increased bus speed to 10MHz, now it bugs out at lower than ~110 microseconds.

**Comment by jonas-merkle** [comment]
Hi Stipena,
I'm sorry, but it's not clear to me what exactly you want to implement within your program. Do I understand it correctly that the motor turns between 10 and 350 degrees back and forth?

First of all, I noticed in your program that you only work with a series of IF statements. I could imagine that in some cases this leads to unwanted side effects and would advise you to use a combination of IF, ELSE IF and ELSE statements.

I assume that you suspect that the command `as5047p.readAngleDegree();` takes too long, so that the timing is no longer correct. Is my assumption correct?

During a read access to the sensor the command `delayMicroseconds(1);` is called at two places in the library. This results in a minimal delay of 2 microseconds. In addition there is the time needed for the actual SPI communication. This time depends on the selected SPI frequency. I think the sensor itself supports SPI frequencies up to 32 MHz. And if I remember correctly an Arduino Uno with 16 MHz clockspeed supports a maximum SPI frequency of only 8 MHz. I can imagine that for your specific application it is simply too slow if the SPI frequency is limited to 8 MHz and you need a more powerful Arduino for your application. I have successfully used both the library and the sensor with SPI frequencies higher than 8 MHz on an Adafruit Feather M0 board and on a Teensy 4.1. But I haven't done any real performance analysis yet, so I can't give any concrete information about the expected delay when reading a measurement. If I find time in the next few days I will do some tests in this regard. 

What you could try is to use the function `as5047p.readAngleRaw();` instead of the function `as5047p.readAngleDegree();`. This should increase the performance minimally, because this function works with integer values only. Processing floating point values on 8-bit AVR microcontrollers is significantly slower than working exclusively with integer values due to the lack of FPU. For your application I don't see any advantage in using floating point values.

I hope that this information helps you a little bit and if you have new insights I would be glad if you pass them on.

Best regards
Jonas

**Comment by jonas-merkle** [comment]
I have now built a small performance test for the library. For this I didn't use an Aruino Uno but a Teensy 4.1.
My results are the following:

| SPI Frequency [in Hz] | Average Delay for 'readAngleRaw()' [in µSec] | Average Delay for 'readAngleRaw()' [in µSec] |
|------------------------------:|-------------------------------------------------------------:|-------------------------------------------------------------:|
| 100000                        | 37                                                                     | 37                                                                     |
| 8000000                      | 7                                                                       | 7                                                                       |
| 20000000                    | 4                                                                       | 4                                                                       |
| 28000000                    | 4                                                                       | 4                                                                       |

However, you have to keep in mind that the Teensy 4.1 is a board with 32-bit ARM Cortex M7 with 600 MHz CPU clock and this chip also has an FPU for floating point operations.

The code of the performance test is available in [this separate repository](https://github.com/jonas-merkle/AS5047P-Library-Performance-Test).

**Comment by Stipena** [comment]
Yes, motor turns between 10 and 350 degrees. Yes, going with angleReadRaw made it work faster but not much. However, I changed from Arduino Uno to Micro, reduced the code to read angle at intervals and spin stepper continuously and reduce step interval after it made 12800 steps. It works well up to certain point. If you try to read angle to frequently then it bugs out (stops responding to some step interval changes). Somewhere around 500us reading interval it doesn't work perfectly but kinda ok. But for me it would be nice to read angle at every step.
I will try Teensy, found one in the drawer, never used one before. If I'm right it is 3.6 version.

**Comment by jonas-merkle** [comment]
I hope that a board with a higher CPU clock frequency will solve your problems!
If you have time and desire I would be grateful if you could run the performance test (see [here](https://github.com/jonas-merkle/AS5047P-Library-Performance-Test/blob/master/src/main.cpp)) with your setup (Arduino Uno / Nano) and share your results. Unfortunately I don't have such an Arduino handy at the moment, but I'm interested in the delay one has to expect when reading the measured value on such boards.

**Comment by Stipena** [comment]
I tried on Uno, Micro, but both give only this
![failed test](https://user-images.githubusercontent.com/4146267/114222266-0771f980-9977-11eb-8573-5f184e61fe89.PNG)

**Comment by jonas-merkle** [comment]
Ohh looks like sprintf is not working as expected again... 🙈

I have adjusted the formatting pattern... Now the results should be displayed also with the Arduino Uno / Nano.

**Comment by Stipena** [comment]
Micro 100 000Hz - 287us, 8 000 000Hz - 35us
Mega2560 100 000Hz - 288us, 8 000 000Hz - 36us
Uno/Nano 100 000Hz - 291us, 8 000 000Hz - 39us

Can you add Teensy connection to description? I'm afraid to fry something, since there is if I'm correct 3 SPI buses. And Teensy CS pin would go to CSn on AS5047P?

**Comment by jonas-merkle** [comment]
> Micro 100 000Hz - 287us, 8 000 000Hz - 35us
> Mega2560 100 000Hz - 288us, 8 000 000Hz - 36us
> Uno/Nano 100 000Hz - 291us, 8 000 000Hz - 39us

Thanks for the effort and your results!

> Can you add Teensy connection to description? I'm afraid to fry something, since there is if I'm correct 3 SPI buses. And Teensy CS pin would go to CSn on AS5047P?

Sure, which version of the Teensy Board are you using? The version 3.6 ?

**Comment by Stipena** [comment]
Yes 3.6

**Comment by jonas-merkle** [comment]
To connect the AS5047P sensor to Teensy 3.6 board follow this pin mapping:

| AS5047P Pin | Teensy Pin | Comment |
|:-----------:|:-----------:|:--------|
| GND | GND | - |
| VDD | 3V | - |
| VDD3V| 3V | - |
| MOSI | MOSI0 (Pin: 11) | - |
| MISO | MISO0 (Pin: 12) | - |
| CLK | SCK0 (Pin: 13) | - |
| CSn | CS0 (Pin: 10) | Can be any unused digital pin on the teenys as long as it's configured here `AS5047P as5047p(<ChipSelectPort>);` |

**Please note that the sensor can no longer be operated with 5V, but must be operated with 3.3V, otherwise there is a risk of damaging the Teensy Board!**

**Comment by jonas-merkle** [comment]
Regarding the actual problem it can be summarized that the expected delay when reading measured values is massively dependent on the used SPI frequency. 
I will test in the future if it is possible to replace the `delayMicroseconds(1);` with a no-op operation to get a minimal performance gain.

**Comment by Stipena** [comment]
Thanks! From Arduino forum, some suggest to try timer/interrupts. So far had no time to learn them as it is a bit hard (I'm not that great at math or programming).

**Comment by Stipena** [comment]
Made setup on Teensy 3.6 with 28MHz bus speed, issue still persists :/

**Comment by jonas-merkle** [comment]
I am still in the process of replacing the `delayMicroseconds()` in the library to get some more speed. However, I can't get below a delay of 3 microseconds when reading the measurements.

Meanwhile I have the feeling that there is another problem in your software besides the delay in reading the readings from the sensor. Or that there is an interference between the other libraries used in your software.
Switching from your approach with `micros()` to the use of timers is in my opinion also a possibility, but I agree with you that it becomes more complex and less easy to understand. I think that in general it is also feasible with `micros()`.

**Comment by Stipena** [comment]
I give you latest code to have Idea what is happening 
```
// include the library for the AS5047P sensor.
#include <AS5047P.h>

// define the chip select port.
#define AS5047P_CHIP_SELECT_PORT 10

// define the spi bus speed
#define AS5047P_CUSTOM_SPI_BUS_SPEED 28000000

// initialize a new AS5047P sensor object.
AS5047P as5047p(AS5047P_CHIP_SELECT_PORT, AS5047P_CUSTOM_SPI_BUS_SPEED);

//***STEPPER***//
////////////////
const int EnableCutterPin = 2;
const int CutterDirPin = 3;
const int CutterStepPin = 4;
int EnableCutterState = HIGH; // Cutter Disabled
int CutterDirState = HIGH; // Cutter spin clockwise
int CutterStepState = LOW; // Cutter step off
unsigned long CutterStepMicros = 0;
unsigned long stepInterval = 100;
uint16_t angle;
unsigned long readingInterval = 10;
unsigned long counter = 0;

// arduino setup routine
void setup() {
  pinMode(EnableCutterPin, OUTPUT); //Enable Cutter
  pinMode(CutterDirPin, OUTPUT); //Cutter Dir
  pinMode(CutterStepPin, OUTPUT); //Cutter Step
  digitalWrite(CutterDirPin, CutterDirState);
  // initialize the serial bus for the communication with your pc.
  Serial.begin(115200);

  // initialize the AS5047P sensor and hold if sensor can't be initialized..
  while (!as5047p.initSPI()) {
    // Serial.println(F("Can't connect to the AS5047P sensor! Please check the connection... "));
    delay(5000);
  }

}

// arduino loop routine
void loop() {
  angle = as5047p.readAngleRaw();      // read the angle value from the AS5047P sensor an print it to the serial consol.
  runStepper();
}

//***RUN STEPPER***//
////////////////////
void runStepper() {
  unsigned long currentMicros = micros();
  if (currentMicros - CutterStepMicros >= stepInterval) {
    CutterStepMicros = currentMicros;
    if (CutterStepState == LOW) {
      CutterStepState = HIGH;
      counter++;
      if (counter == 6400) {
        counter = 0;
        stepInterval--;
        Serial.println(stepInterval);
        if (stepInterval <= 30) {
          stepInterval = 100;
        }
      }
    } else {
      CutterStepState = LOW;
    }
    digitalWrite(CutterStepPin, CutterStepState);
  }
}
```
With this code stepper should increase speed by 1us, but I literally can hear from stepper that it doesn't do that. Only changes sound every 3rd, 4th or 5th increase. Point at which it changes speed, seems depend on `stepInterval` value.
Only AS5047P library is included in code.

**Comment by jonas-merkle** [comment]
I have written the following [program](https://gist.github.com/jonas-merkle/767becf922b0faf8d1c69d7a5ba3b4aa). It reads the AS5047P sensor and at the same time generates a pulse sequence to control the stepper motor driver.

I have an A4988 chip but unfortunately no suitable stepper motor on which I could test the software. Therefore I looked at the generated signal with a logic analyzer. The attached picture shows the signal for a step interval length of 12 microseconds.

![image](https://user-images.githubusercontent.com/23365138/114269454-7860f200-9a07-11eb-8ad5-3b07ec8124eb.png)

**Comment by Stipena** [comment]
Had to tweak it to work:
```
// read the AS5047P sensor if the enable flag is set to true
	if (AS5047P_ENABLED == true)
	{
		uint16_t angle = as5047p.readAngleRaw();

		// change the direction
		if (angle > 1000 && DirState == LOW)
		{
			DirState = HIGH;
		}
		else if (angle < 100 && DirState == HIGH)
		{
			DirState = LOW;
		}

		digitalWrite(DIR_PIN,   DirState);
	}
```
Now it runs back and forth, instead of spinning continuously.
Anyhow, now I got my hands on FastAccelStepper library which uses timers. Will do some testing next week.

**Comment by Stipena** [comment]
Can confirm, FastAccelStepper library works fine parallel with AS5047P library. On Arduino Nano achieves speeds of 25000steps/s.
